docs/rfcs/0005-json-library.md

```diff
@@ -0,0 +1,48 @@
+---
+layout: page
+title: RFC process
+permalink: /rfc/0005-json-library
+---
+
+* Start date: 2020-07-28
+* RFC PR: TBD
+
+## Summary
+
+Introduce a library for reading and writing JSON into Android Components and Fenix.
+
+## Motivation
+
+We currently interact with JSON in many areas of Android Components and Fenix. Currently serialization and deserialization is all done manually with `JSONObject`, which creates more work for developers and potential bugs. Fenix also is interested in using a library and we'd like to avoid shipping multiple JSON libraries to users. Using a JSON library allows us to reduce code that needs to be written, potentially gain some performance wins from a well maintained lib, and avoid bugs in our custom JSON implementations.
+
+Some examples of JSON usage today are:
+- Reading a `WebAppManifest`
+- Parsing API calls in various service components
+- Sending messages to and from extensions
+
+## Guide-level explanation
+
+We'll start with selecting a JSON library to integrate into Android Components and Fenix. Some options we've looked at include:
+- [kotlinx-serialization](https://github.com/Kotlin/kotlinx.serialization): Multi-platform, multi-format, built by JetBrains. Also much newer than other solutions so documentation is relatively lacking.
+- [Gson](https://github.com/google/gson): Long standing library built by Google. Slower than other options.
+- [Moshi](https://github.com/square/moshi): Simpler API, built by Square. Fewer features than Gson but also faster.
+
+Once a library is chosen we can use it when building new components that use JSON and slowly migrate existing code.
+
+## Drawbacks
+
+* Using a new tool means we have to invest some time in learning how to use it. We're at least familiar with JSONObject as are most people. This "having another thing to learn" thing is especially problematic if Fenix and other apps adopt different technologies to fill the same gap.
+* We want to be conservative adding third-party libraries.
+
+## Rationale and alternatives
+
+We currently use `JSONObject` for our JSON decoding and encoding. This forces duplication of logic (need to functions for both String -> JSON and JSON -> String) and creates potential crashes if we don't catch JSON exceptions properly. A well maintained library allows us to write less code and potentially gain performance wins.
```

Comment:
I agree that using a library means much less code and comfort for AC developers. From my experience with GSON it is also possible to have the same kind of crashes if you end up with incompatible Kotlin models and JSON. On that front I wonder how well those libraries can deal with versioned/flexible formats? They all look great in isolated examples - how do they deal with changing formats (forward and backward compatibility)?

Reply:
kotlinx-serialization is able to handle optional/required fields, and different schemas could be represented by different classes if needed.